File: JSONPlaces/PlacesViewController.m
//
//  PlacesViewController.m
//  JSONPlaces
//
//  Created by Adrian Kubała on 13.02.2017.
//  Copyright © 2017 Adrian. All rights reserved.
//

#import "PlacesViewController.h"
#import "JSONParser.h"
#import "NSURL+Image.h"
#import "UIImage+Resizing.h"
#import "Place.h"
#import "PlaceView.h"
#import "CLLocation+Distance.h"
#import "NearbyPlacesViewController.h"

@interface PlacesViewController ()

@end

@implementation PlacesViewController

- (void)viewDidLoad {
  [super viewDidLoad];
  
  [self setupArrays];
  [self setupParser];
}

- (void)setupParser {
  NSURL *jsonURL = [[NSURL alloc] initWithString:@"http://pastebin.com/raw/dTYu3jmN"];
  
  JSONParser *jsonParser = [[JSONParser alloc] initWithURL:jsonURL];
  [jsonParser fetch:^(NSArray *result, NSError *error) {
    if (error) {
      NSLog(@"%@", error.localizedDescription);
      return;
    }
    
    if (result) {
      [self fillPlacesWithJSON:result];
    }
  }];
}

- (void)setupArrays {
  self.places = [[NSMutableArray alloc] init];
  self.matchedPlaces = [[NSMutableArray alloc] init];
}

- (void)fillPlacesWithJSON:(NSArray *)jsonArray {
  for (NSDictionary *object in jsonArray) {
    NSString *name = object[@"name"];
    NSURL *imageURL = [[NSURL alloc] initWithString:object[@"pin_url"]];
    
    NSDictionary *coordinate = [[NSDictionary alloc] initWithDictionary:object[@"coordinate"]];
    double lat = [[coordinate objectForKey:@"latitude"] doubleValue];
    double lon = [[coordinate objectForKey:@"longitude"] doubleValue];
    
    [imageURL getImage:^(UIImage * image) {
      [image resizeWithNewWidth:30 completion:^(UIImage * scaledImage) {
        
        Place *place = [[Place alloc] initWithName:name latitude:lat longitude:lon pinImage:scaledImage];
        [self.places addObject:place];
        
        [self.tableView reloadData];
      }];
    }];
  }
}

#pragma mark - Table view data source

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
  return self.places.count;
}


- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
  PlaceView *cell = (PlaceView *) [tableView dequeueReusableCellWithIdentifier:@"PlaceView" forIndexPath:indexPath];
  
  Place *place = self.places[indexPath.row];
  cell.label.text = place.name;
  cell.pinImageView.image = place.pinImage;
  
  return cell;
}

# pragma mark UITableViewDelegate

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
  [self.matchedPlaces removeAllObjects];
  
  Place *place = self.places[indexPath.row];
  CLLocation *location = place.location;
  [self getAllDistancesFrom:location atCellRow:indexPath.row];
  
  [self performSegueWithIdentifier:@"ShowNearbyPlaces" sender:self];
  
}

- (void)getAllDistancesFrom:(CLLocation *)location atCellRow:(NSInteger)cellRow {
  for (Place *place in self.places) {
    NSUInteger i = [self.places indexOfObject:place];
    
    if (i == cellRow) {
      continue;
    }
    
    CLLocation * otherLocation = place.location;
    double distance = [location distanceInKmTo:otherLocation];
    
    if (distance <= 2) {
      Place *matchedPlace = place;
      matchedPlace.distance = distance;
      [self.matchedPlaces addObject:matchedPlace];
    }
  }
}

- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
  NearbyPlacesViewController *nearbyPlacesViewController = (NearbyPlacesViewController *) segue.destinationViewController;
  nearbyPlacesViewController.nearbyPlaces = [[NSMutableArray alloc] initWithArray:self.matchedPlaces];
}

@end

File: JSONPlaces/PlacesViewController.swift
//
//  PlacesViewController.swift
//  JSONPlaces
//
//  Created by Adrian on 09.08.2016.
//  Copyright © 2016 Adrian. All rights reserved.
//

import UIKit

class PlacesViewController: UITableViewController {
  var places: [Place] = []
  var matchedPlaces: [Place] = []
  
  override func viewDidLoad() {
    super.viewDidLoad()
    
    let jsonURL = URL(string: "http://pastebin.com/raw/dTYu3jmN")
    
    let jsonParser = JSONParser(url: jsonURL!)
    jsonParser?.fetch { [weak self] (result, error) in
      if let error = error {
        print(error.localizedDescription)
        return
      }
      
      if let result = result {
        self?.fillPlacesWithJSON(result)
      }
    }
  }
  
  func fillPlacesWithJSON(_ jsonArray: [Any]) {
    for object in jsonArray {
      let entry = object as! [String: Any]
      
      let name = entry["name"] as! String
      let imageURL = NSURL(string: entry["pin_url"] as! String)
      
      let coordinate = entry["coordinate"] as! [String: String]
      let lat = Double(coordinate["latitude"]!)
      let lon = Double(coordinate["longitude"]!)
      
      imageURL?.getImage { (image) in
        image?.resize(withNewWidth: 30) { (scaledImage) in
          
          let place = Place(name: name, latitude: lat!, longitude: lon!, pinImage: scaledImage)
          self.places.append(place!)
          
          self.tableView.reloadData()
        }  
      }
    }
  }
  
  override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
    return places.count
  }
  
  override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
    let cell = tableView.dequeueReusableCell(withIdentifier: "PlaceView", for: indexPath) as! PlaceView
    
    cell.label.text = places[indexPath.row].name
    cell.pinImageView.image = places[indexPath.row].pinImage
    
    return cell
  }
  
  override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
    matchedPlaces = []
    
    let location = places[indexPath.row].location
    getAllDistancesFrom(location: location!, cellRow: indexPath.row)
    
    performSegue(withIdentifier: "ShowNearbyPlaces", sender: self)
  }
  
  func getAllDistancesFrom(location: CLLocation, cellRow: Int) {
    for (i, place) in places.enumerated() {
      if i == cellRow {
        continue
      }
      
      let otherLocation = place.location
      let distance = location.distanceInKm(to: otherLocation)
      
      if distance <= 2 {
        let matchedPlace = place
        matchedPlace.distance = distance
        matchedPlaces.append(matchedPlace)
      }
    }
  }
  
  override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
    let nearbyPlacesViewController = segue.destination as! NearbyPlacesViewController
    nearbyPlacesViewController.nearbyPlaces = matchedPlaces
  }
}
